src/librustc/ty/sty.rs

```diff
@@ -372,31 +366,33 @@ pub struct ClosureSubsts<'tcx> {
 struct SplitClosureSubsts<'tcx> {
     closure_kind_ty: Ty<'tcx>,
     closure_sig_ty: Ty<'tcx>,
-    upvar_kinds: &'tcx [GenericArg<'tcx>],
+    tupled_upvars_ty: Ty<'tcx>,
 }
 
 impl<'tcx> ClosureSubsts<'tcx> {
     /// Divides the closure substs into their respective
     /// components. Single source of truth with respect to the
     /// ordering.
-    fn split(self, def_id: DefId, tcx: TyCtxt<'_>) -> SplitClosureSubsts<'tcx> {
-        let generics = tcx.generics_of(def_id);
-        let parent_len = generics.parent_count;
+    fn split(self) -> SplitClosureSubsts<'tcx> {
```

Comment:
I've already mentioned this to @eddyb via PM, but it could be useful for `SplitClosureSubsts` to contain a `parent_substs: &'tcx [GenericArg<'tcx>]` field - that way any callers that need to access the specific components of the closure substs (and use `SplitClosureSubsts` to do so) can also access the parent substs without having to reproduce some of this logic.

Reply:
You can add this in your PR, FWIW, but if @nikomatsakis doesn't mind I can do it in this one too.